File: dungeons-1.21.11/src/main/java/fishmod/utils/SkyblockItems.java
package fishmod.utils;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Resolves display names → Skyblock item IDs by fetching the canonical list from
 * https://api.hypixel.net/v2/resources/skyblock/items at startup.
 * Falls back silently to a null/empty map if the fetch fails — callers should keep their
 * hardcoded fallback maps for that case.
 */
public class SkyblockItems {
    private static final HttpClient HTTP = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(10)).build();
    private static final Map<String, String> nameToId = new HashMap<>();
    private static final Map<String, Double> npcSellPrice = new HashMap<>();
    private static final AtomicBoolean loading = new AtomicBoolean(false);
    private static volatile boolean loaded = false;

    public static void initAsync() {
        if (loaded || !loading.compareAndSet(false, true)) return;
        HttpRequest req = HttpRequest.newBuilder()
                .uri(URI.create("https://api.hypixel.net/v2/resources/skyblock/items"))
                .timeout(Duration.ofSeconds(15))
                .GET().build();
        HTTP.sendAsync(req, HttpResponse.BodyHandlers.ofString())
                .thenAccept(r -> {
                    try {
                        if (r.statusCode() != 200) return;
                        JsonObject obj = JsonParser.parseString(r.body()).getAsJsonObject();
                        JsonArray items = obj.getAsJsonArray("items");
                        Map<String, String> nextNames = new HashMap<>();
                        Map<String, Double> nextNpc = new HashMap<>();
                        for (int i = 0; i < items.size(); i++) {
                            JsonObject it = items.get(i).getAsJsonObject();
                            if (!it.has("id") || !it.has("name")) continue;
                            String id = it.get("id").getAsString();
                            String name = it.get("name").getAsString().replaceAll("§.", "").trim();
                            if (name.isEmpty()) continue;
                            nextNames.put(name, id);
                            if (it.has("npc_sell_price")) {
                                try { nextNpc.put(id, it.get("npc_sell_price").getAsDouble()); } catch (Exception ignored) {}
                            }
                        }
                        synchronized (nameToId) { nameToId.clear(); nameToId.putAll(nextNames); }
                        synchronized (npcSellPrice) { npcSellPrice.clear(); npcSellPrice.putAll(nextNpc); }
                        loaded = true;
                        fishmod.utils.debug.Debug.LOGGER.info("[SkyblockItems] loaded {} items ({} with NPC sell price)", nextNames.size(), nextNpc.size());
                        // Re-apply price mode in case CroesusPrices already loaded the bazaar
                        fishmod.features.croesus.CroesusPrices.applyPriceMode();
                    } catch (Exception ignored) {
                    } finally {
                        loading.set(false);
                    }
                })
                .exceptionally(t -> { loading.set(false); return null; });
    }

    /**
     * Case-insensitive name search for autocomplete. Prefix matches rank before substring
     * matches; both are sorted alphabetically and the combined result is capped at {@code limit}.
     * Returns display names (resolve to ids via {@link #idFor(String)}).
     */
    public static java.util.List<String> searchNames(String query, int limit) {
        if (query == null || query.isBlank()) return java.util.List.of();
        String q = query.toLowerCase();
        java.util.List<String> prefix = new java.util.ArrayList<>();
        java.util.List<String> contains = new java.util.ArrayList<>();
        synchronized (nameToId) {
            for (String name : nameToId.keySet()) {
                String ln = name.toLowerCase();
                if (ln.startsWith(q)) prefix.add(name);
                else if (ln.contains(q)) contains.add(name);
            }
        }
        prefix.sort(String.CASE_INSENSITIVE_ORDER);
        contains.sort(String.CASE_INSENSITIVE_ORDER);
        java.util.List<String> out = new java.util.ArrayList<>(prefix);
        for (String c : contains) {
            if (out.size() >= limit) break;
            out.add(c);
        }
        return out.size() > limit ? new java.util.ArrayList<>(out.subList(0, limit)) : out;
    }

    /** @return canonical id for the given display name, or null if not loaded/unknown */
    public static String idFor(String name) {
        if (name == null) return null;
        synchronized (nameToId) {
            return nameToId.get(name);
        }
    }

    public static boolean isLoaded() { return loaded; }

    public static double npcSellPriceFor(String id) {
        if (id == null) return 0;
        synchronized (npcSellPrice) {
            Double v = npcSellPrice.get(id);
            return v == null ? 0 : v;
        }
    }
    public static Map<String, Double> npcSellPriceMap() {
        synchronized (npcSellPrice) { return new HashMap<>(npcSellPrice); }
    }
}

File: dungeons-1.21.11/src/main/kotlin/fishmod/utils/SkyblockItems.kt
package fishmod.utils

import com.google.gson.JsonParser
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Resolves display names -> Skyblock item IDs by fetching the canonical list from
 * https://api.hypixel.net/v2/resources/skyblock/items at startup.
 * Falls back silently to a null/empty map if the fetch fails — callers should keep their
 * hardcoded fallback maps for that case.
 */
object SkyblockItems {
    private val HTTP: HttpClient = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(10)).build()
    private val nameToId = HashMap<String, String>()
    private val npcSellPrice = HashMap<String, Double>()
    private val loading = AtomicBoolean(false)
    @Volatile
    private var loaded = false

    @JvmStatic
    fun initAsync() {
        if (loaded || !loading.compareAndSet(false, true)) return
        val req = HttpRequest.newBuilder()
            .uri(URI.create("https://api.hypixel.net/v2/resources/skyblock/items"))
            .timeout(Duration.ofSeconds(15))
            .GET().build()
        HTTP.sendAsync(req, HttpResponse.BodyHandlers.ofString())
            .thenAccept { r ->
                try {
                    if (r.statusCode() != 200) return@thenAccept
                    val obj = JsonParser.parseString(r.body()).asJsonObject
                    val items = obj.getAsJsonArray("items")
                    val nextNames = HashMap<String, String>()
                    val nextNpc = HashMap<String, Double>()
                    for (i in 0 until items.size()) {
                        val it = items[i].asJsonObject
                        if (!it.has("id") || !it.has("name")) continue
                        val id = it.get("id").asString
                        val name = it.get("name").asString.replace(Regex("§."), "").trim()
                        if (name.isEmpty()) continue
                        nextNames[name] = id
                        if (it.has("npc_sell_price")) {
                            try {
                                nextNpc[id] = it.get("npc_sell_price").asDouble
                            } catch (ignored: Exception) {
                            }
                        }
                    }
                    synchronized(nameToId) { nameToId.clear(); nameToId.putAll(nextNames) }
                    synchronized(npcSellPrice) { npcSellPrice.clear(); npcSellPrice.putAll(nextNpc) }
                    loaded = true
                    fishmod.utils.debug.Debug.LOGGER.info("[SkyblockItems] loaded {} items ({} with NPC sell price)", nextNames.size, nextNpc.size)
                    // Re-apply price mode in case CroesusPrices already loaded the bazaar
                    fishmod.features.croesus.CroesusPrices.applyPriceMode()
                } catch (ignored: Exception) {
                } finally {
                    loading.set(false)
                }
            }
            .exceptionally { loading.set(false); null }
    }

    /**
     * Case-insensitive name search for autocomplete. Prefix matches rank before substring
     * matches; both are sorted alphabetically and the combined result is capped at `limit`.
     * Returns display names (resolve to ids via [idFor]).
     */
    @JvmStatic
    fun searchNames(query: String?, limit: Int): List<String> {
        if (query == null || query.isBlank()) return listOf()
        val q = query.lowercase()
        val prefix = ArrayList<String>()
        val contains = ArrayList<String>()
        synchronized(nameToId) {
            for (name in nameToId.keys) {
                val ln = name.lowercase()
                if (ln.startsWith(q)) prefix.add(name)
                else if (ln.contains(q)) contains.add(name)
            }
        }
        prefix.sortWith(String.CASE_INSENSITIVE_ORDER)
        contains.sortWith(String.CASE_INSENSITIVE_ORDER)
        val out = ArrayList<String>(prefix)
        for (c in contains) {
            if (out.size >= limit) break
            out.add(c)
        }
        return if (out.size > limit) ArrayList(out.subList(0, limit)) else out
    }

    /** @return canonical id for the given display name, or null if not loaded/unknown */
    @JvmStatic
    fun idFor(name: String?): String? {
        if (name == null) return null
        synchronized(nameToId) {
            return nameToId[name]
        }
    }

    @JvmStatic
    fun isLoaded(): Boolean = loaded

    @JvmStatic
    fun npcSellPriceFor(id: String?): Double {
        if (id == null) return 0.0
        synchronized(npcSellPrice) {
            val v = npcSellPrice[id]
            return v ?: 0.0
        }
    }

    @JvmStatic
    fun npcSellPriceMap(): Map<String, Double> {
        synchronized(npcSellPrice) { return HashMap(npcSellPrice) }
    }
}
